source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1237 @@
+        {
+          grid_extent[i].first = geometry->get_origin()[i];
+          grid_extent[i].second = geometry->get_extents()[i];
+        }
+
+      // TODO: There has to be a better type to use to get this.
+      // const unsigned int repetitions[dim] = {geometry->get_repetitions()};
+      const unsigned int x_repetitions = geometry->get_repetitions(0);
+      const unsigned int y_repetitions = geometry->get_repetitions(1);
+
+      // Set nx and dx, as these will be the same regardless of dimension.
+      nx = 1+(2*use_ghost)+std::pow(2,surface_resolution+additional_refinement)*x_repetitions;
+      dx = (grid_extent[0].second - grid_extent[0].first)/(nx-1-(2*use_ghost));
+      x_extent = (grid_extent[0].second - grid_extent[0].first)+2*dx*use_ghost;
+
+      // Sub intervals are 3 less than points, if including the ghost nodes. Otherwise 1 less.
+      table_intervals[0] = nx-1-(2*use_ghost);
+      // TODO: it'd be best to not have to use dim-1 intervals at all.
+      table_intervals[dim-1] = 1;
+
+      if (dim == 2)
+        {
+          dy = dx;
+          y_extent = round(y_extent_2d/dy)*dy+2*dy*use_ghost;
+          ny = 1+y_extent/dy;
+        }
+      else
+        {
+          ny = 1+(2*use_ghost)+std::pow(2,surface_resolution+additional_refinement)*y_repetitions;
+          dy = (grid_extent[1].second - grid_extent[1].first)/(ny-1-(2*use_ghost));
+          table_intervals[1] = ny-1-(2*use_ghost);
+          y_extent = (grid_extent[1].second - grid_extent[1].first)+2*dy*use_ghost;
+        }
+
+      // Determine array size to send to fastscape
+      array_size = nx*ny;
+
+      // Create a folder for the FastScape visualization files.
+      Utilities::create_directory (this->get_output_directory() + "VTK/",
+                                   this->get_mpi_communicator(),
+                                   false);
+      last_output_time = 0;
+    }
+
+
+    template <int dim>
+    void
+    FastScape<dim>::compute_velocity_constraints_on_boundary(const DoFHandler<dim> &mesh_deformation_dof_handler,
+                                                             AffineConstraints<double> &mesh_velocity_constraints,
+                                                             const std::set<types::boundary_id> &boundary_ids) const
+    {
+      TimerOutput::Scope timer_section(this->get_computing_timer(), "Fastscape plugin");
+      const types::boundary_id relevant_boundary = this->get_geometry_model().translate_symbolic_boundary_name_to_id ("top");
+      const int current_timestep = this->get_timestep_number ();
+
+      double a_dt = this->get_timestep();
+      if (this->convert_output_to_years())
+        {
+          a_dt = this->get_timestep()/year_in_seconds;
+        }
+
+      // We only want to run fastscape if there was a change in time.
+      if (a_dt > 0)
+        {
+          /*
+           * Initialize a vector of temporary variables to hold: z component, index, Vx, Vy, and Vz.
```

Comment:
Can you explain why? The fact that a vector of tmp vars is initialized can be seen from the code, why it's necessary not.